File: backend/app/routes/yarn_adapter.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
import asyncio
import random

router = APIRouter()

class YarnQueryRequest(BaseModel):
    text: str
    language: str = "en-ng"
    speaker: str = "default"

class YarnResponse(BaseModel):
    text: str
    audio_url: str
    language: str
    processing_time: float

# Demo responses for YarnGPT - deterministic based on input
DEMO_RESPONSES = {
    # Nigerian English responses
    "en-ng": {
        "default": {
            "text": "I understand you want to know more about your data privacy. In Nigeria, organizations must get your clear consent before using your personal data. You have the right to know what data they collect, why they need it, and how long they keep it.",
            "audio_file": "en_ng_default.mp3"
        },
        "explain_access": {
            "text": "When an organization accesses your data, they should have a valid reason that you previously agreed to. For example, banks access your transaction history to verify your identity for loans, which helps prevent fraud and ensures you qualify for the right financial products.",
            "audio_file": "en_ng_explain.mp3"
        },
        "consent_rights": {
            "text": "Your consent rights in Nigeria include: the right to be informed about data collection, the right to access your data, the right to correct wrong information, the right to delete your data, and the right to withdraw consent at any time.",
            "audio_file": "en_ng_rights.mp3"
        },
        "banking_data": {
            "text": "Banks need your financial data to assess your creditworthiness and comply with Central Bank of Nigeria regulations. They use your transaction history to understand your spending patterns and determine if you can repay loans safely.",
            "audio_file": "en_ng_banking.mp3"
        },
        "telecom_data": {
            "text": "Telecommunications companies like MTN collect your usage data to improve network quality and provide better services. They also use location data to optimize network coverage in your area.",
            "audio_file": "en_ng_telecom.mp3"
        }
    },
    
    # Igbo responses
    "ig": {
        "default": {
            "text": "Aghọtara m na ịchọrọ ịmata ihe gbasara nchekwa data gị. Na Naịjirịa, ụlọ ọrụ ga-enweta nkwenye gị doro anya tupu ha eji data nkeonwe gị mee ihe.",
            "audio_file": "ig_default.mp3"
        },
        "explain_access": {
            "text": "Mgbe ụlọ ọrụ na-enweta data gị, ha kwesịrị inwe ezigbo ihe kpatara ya nke ị kwenyere na mbụ. Dịka ọmụmaatụ, ụlọ akụ na-elele akụkọ ego gị iji chọpụta onye ị bụ maka mbinye ego.",
            "audio_file": "ig_explain.mp3"
        },
        "consent_rights": {
            "text": "Ikike nkwenye gị na Naịjirịa gụnyere: ikike ịmata mgbe a na-anakọta data gị, ikike ịnweta data gị, ikike imezi ozi ezighi ezi, ikike ihichapụ data gị, na ikike ịkwụsị nkwenye mgbe ọ bụla.",
            "audio_file": "ig_rights.mp3"
        }
    },
    
    # Yoruba responses
    "yo": {
        "default": {
            "text": "Mo ye mi pe o fe mo nipa aabo data re. Ni Nigeria, awon ile-ise gbodo gba igbanilaaye to han kedere lati odo re ki won to lo data ti ara re.",
            "audio_file": "yo_default.mp3"
        },
        "explain_access": {
            "text": "Nigbati ile-ise kan ba n wole si data re, won gbodo ni idi to dara ti o ti gba lati tele. Fun apere, awon ile-owo n wo itan owo re lati ri daju pe eni ti o je fun awin owo.",
            "audio_file": "yo_explain.mp3"
        },
        "consent_rights": {
            "text": "Awon eto re nipa igbanilaaye ni Nigeria ni: eto lati mo nigbati won ba n gba data re, eto lati ri data re, eto lati se atunse alaye ti ko tọ, eto lati pa data re rẹ, ati eto lati fa igbanilaaye re pada nigbakugba.",
            "audio_file": "yo_rights.mp3"
        }
    },
    
    # Hausa responses
    "ha": {
        "default": {
            "text": "Na fahimci kana son ka san game da kare bayananku. A Najeriya, kamfanoni dole su sami izininku da bayyane kafin su yi amfani da bayananku na sirri.",
            "audio_file": "ha_default.mp3"
        },
        "explain_access": {
            "text": "Lokacin da kamfani ya shiga bayananku, ya kamata su sami dalili mai kyau da kuka yarda da shi a baya. Misali, bankuna suna kallon tarihin kuɗin ku don tabbatar da ko kun cancanci rancen da kuke nema.",
            "audio_file": "ha_explain.mp3"
        },
        "consent_rights": {
            "text": "Hakokinku na yarda a Najeriya sun haɗa da: hakin sanin lokacin da ake tattara bayananku, hakin samun bayananku, hakin gyara bayanan da ba daidai ba, hakin share bayananku, da hakin janye yardan ku kowane lokaci.",
            "audio_file": "ha_rights.mp3"
        }
    }
}

def _determine_response_type(text: str) -> str:
    """Determine response type based on input text keywords"""
    text_lower = text.lower()
    
    if any(keyword in text_lower for keyword in ['explain', 'why', 'access', 'purpose', 'first bank', 'bank']):
        if any(keyword in text_lower for keyword in ['bank', 'first bank', 'financial', 'loan', 'transaction']):
            return 'banking_data'
        elif any(keyword in text_lower for keyword in ['mtn', 'telecom', 'phone', 'network', 'usage']):
            return 'telecom_data'
        else:
            return 'explain_access'
    elif any(keyword in text_lower for keyword in ['rights', 'consent', 'withdraw', 'delete', 'revoke']):
        return 'consent_rights'
    else:
        return 'default'

@router.post("/query", response_model=YarnResponse)
async def yarn_query(request: YarnQueryRequest):
    """
    Demo YarnGPT query endpoint - returns deterministic responses
    
    In production, this would:
    1. Send request to YarnGPT inference server
    2. Process audio generation with WavTokenizer
    3. Return generated audio file URL
    """
    
    # Simulate processing time
    processing_start = asyncio.get_event_loop().time()
    await asyncio.sleep(random.uniform(0.8, 2.0))  # Realistic API delay
    
    # Get language responses
    language_responses = DEMO_RESPONSES.get(request.language, DEMO_RESPONSES["en-ng"])
    
    # Determine response type based on input
    response_type = _determine_response_type(request.text)
    
    # Get appropriate response
    response_data = language_responses.get(response_type, language_responses["default"])
    
    processing_time = asyncio.get_event_loop().time() - processing_start
    
    return YarnResponse(
        text=response_data["text"],
        audio_url=f"demo_audio/{response_data['audio_file']}",
        language=request.language,
        processing_time=processing_time
    )

@router.get("/languages")
async def get_supported_languages():
    """Get list of supported languages"""
    return {
        "languages": [
            {
                "code": "en-ng",
                "name": "Nigerian English",
                "native_name": "Nigerian English"
            },
            {
                "code": "ig",
                "name": "Igbo",
                "native_name": "Igbo"
            },
            {
                "code": "yo",
                "name": "Yoruba",
                "native_name": "Yorùbá"
            },
            {
                "code": "ha",
                "name": "Hausa",
                "native_name": "Hausa"
            }
        ],
        "default": "en-ng"
    }

@router.get("/demo/prompts")
async def get_demo_prompts():
    """Get list of demo prompts and their expected responses"""
    return {
        "prompts": [
            {
                "text": "Why did First Bank Nigeria access my transaction history?",
                "expected_type": "banking_data",
                "languages": ["en-ng", "ig", "yo", "ha"]
            },
            {
                "text": "Explain why MTN accessed my usage data",
                "expected_type": "telecom_data",
                "languages": ["en-ng", "ig", "yo", "ha"]
            },
            {
                "text": "What are my consent rights in Nigeria?",
                "expected_type": "consent_rights",
                "languages": ["en-ng", "ig", "yo", "ha"]
            },
            {
                "text": "Tell me about data privacy",
                "expected_type": "default",
                "languages": ["en-ng", "ig", "yo", "ha"]
            }
        ]
    }

"""
REAL YARNGPT INTEGRATION PLACEHOLDER

To replace this demo adapter with the real YarnGPT model:

1. YarnGPT Server Setup:
   - Clone YarnGPT repository from GitHub
   - Install dependencies: torch, transformers, torchaudio, etc.
   - Download WavTokenizer config and pre-trained checkpoint files
   - Configure tokenizer path to point to downloaded model
   - Set up GPU environment (CUDA recommended)

2. Server Implementation:
   - Create FastAPI server wrapping YarnGPT inference
   - Implement POST /generate_audio endpoint
   - Handle text-to-speech conversion with language support
   - Return generated audio file URLs

3. Expected API Interface:
   POST /generate_audio
   {
     "text": "Input text to convert to speech",
     "language": "en-ng|ig|yo|ha",
     "speaker": "default",
     "temperature": 0.7,
     "max_length": 1024
   }
   
   Response:
   {
     "text": "Processed text",
     "audio_url": "http://server/generated/audio_file.wav",
     "language": "en-ng",
     "duration": 5.2,
     "processing_time": 2.1
   }

4. Replacement Steps:
   - Update yarn_query function to make HTTP requests to YarnGPT server
   - Handle audio file storage and URL generation
   - Add error handling for server unavailability
   - Implement caching for repeated queries
   - Add authentication if required

5. Production Considerations:
   - Use GPU-enabled hosting (AWS P3, Google Cloud GPU, etc.)
   - Implement request queuing for high load
   - Add monitoring and logging
   - Set up audio file cleanup/rotation
   - Consider using CDN for audio file delivery

Example replacement code:
```python
import httpx
import asyncio

async def yarn_query_real(request: YarnQueryRequest):
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{YARNGPT_SERVER_URL}/generate_audio",
            json={
                "text": request.text,
                "language": request.language,
                "speaker": request.speaker
            },
            timeout=30.0
        )
        response.raise_for_status()
        return YarnResponse(**response.json())
```
"""

File: frontend/lib/services/yarn_adapter.dart
import 'dart:math';

class YarnAdapter {
  // Demo mode deterministic responses
  static const Map<String, Map<String, dynamic>> _demoResponses = {
    // Nigerian English responses
    'en-ng': {
      'default': {
        'text': 'I understand you want to know more about your data privacy. In Nigeria, organizations must get your clear consent before using your personal data. You have the right to know what data they collect, why they need it, and how long they keep it.',
        'audioUrl': 'demo_audio/en_ng_default.mp3',
      },
      'explain_access': {
        'text': 'When an organization accesses your data, they should have a valid reason that you previously agreed to. For example, banks access your transaction history to verify your identity for loans, which helps prevent fraud and ensures you qualify for the right financial products.',
        'audioUrl': 'demo_audio/en_ng_explain.mp3',
      },
      'consent_rights': {
        'text': 'Your consent rights in Nigeria include: the right to be informed about data collection, the right to access your data, the right to correct wrong information, the right to delete your data, and the right to withdraw consent at any time.',
        'audioUrl': 'demo_audio/en_ng_rights.mp3',
      },
    },
    
    // Igbo responses
    'ig': {
      'default': {
        'text': 'Aghọtara m na ịchọrọ ịmata ihe gbasara nchekwa data gị. Na Naịjirịa, ụlọ ọrụ ga-enweta nkwenye gị doro anya tupu ha eji data nkeonwe gị mee ihe.',
        'audioUrl': 'demo_audio/ig_default.mp3',
      },
      'explain_access': {
        'text': 'Mgbe ụlọ ọrụ na-enweta data gị, ha kwesịrị inwe ezigbo ihe kpatara ya nke ị kwenyere na mbụ. Dịka ọmụmaatụ, ụlọ akụ na-elele akụkọ ego gị iji chọpụta onye ị bụ maka mbinye ego.',
        'audioUrl': 'demo_audio/ig_explain.mp3',
      },
    },
    
    // Yoruba responses
    'yo': {
      'default': {
        'text': 'Mo ye mi pe o fe mo nipa aabo data re. Ni Nigeria, awon ile-ise gbodo gba igbanilaaye to han kedere lati odo re ki won to lo data ti ara re.',
        'audioUrl': 'demo_audio/yo_default.mp3',
      },
      'explain_access': {
        'text': 'Nigbati ile-ise kan ba n wole si data re, won gbodo ni idi to dara ti o ti gba lati tele. Fun apere, awon ile-owo n wo itan owo re lati ri daju pe eni ti o je fun awin owo.',
        'audioUrl': 'demo_audio/yo_explain.mp3',
      },
    },
    
    // Hausa responses
    'ha': {
      'default': {
        'text': 'Na fahimci kana son ka san game da kare bayananku. A Najeriya, kamfanoni dole su sami izininku da bayyane kafin su yi amfani da bayananku na sirri.',
        'audioUrl': 'demo_audio/ha_default.mp3',
      },
      'explain_access': {
        'text': 'Lokacin da kamfani ya shiga bayananku, ya kamata su sami dalili mai kyau da kuka yarda da shi a baya. Misali, bankuna suna kallon tarihin kuɗin ku don tabbatar da ko kun cancanci rancen da kuke nema.',
        'audioUrl': 'demo_audio/ha_explain.mp3',
      },
    },
  };

  /// Generate response in demo mode with deterministic outputs
  Future<Map<String, dynamic>> generateResponse({
    required String text,
    required String language,
  }) async {
    // Simulate API delay
    await Future.delayed(Duration(milliseconds: 800 + Random().nextInt(1200)));
    
    final responses = _demoResponses[language] ?? _demoResponses['en-ng']!;
    
    // Determine response type based on input text
    String responseKey = 'default';
    
    if (_containsKeywords(text, ['explain', 'why', 'access', 'purpose'])) {
      responseKey = 'explain_access';
    } else if (_containsKeywords(text, ['rights', 'consent', 'withdraw', 'delete'])) {
      responseKey = 'consent_rights';
    }
    
    final response = responses[responseKey] ?? responses['default']!;
    
    return {
      'text': response['text'],
      'audioUrl': response['audioUrl'],
    };
  }

  bool _containsKeywords(String text, List<String> keywords) {
    final lowerText = text.toLowerCase();
    return keywords.any((keyword) => lowerText.contains(keyword));
  }

  /* 
   * REAL YARNGPT ADAPTER PLACEHOLDER
   * 
   * To replace this demo adapter with the real YarnGPT model:
   * 
   * 1. Set up YarnGPT server:
   *    - Clone the YarnGPT repository
   *    - Install Python dependencies (torch, transformers, etc.)
   *    - Download WavTokenizer config and checkpoint files
   *    - Configure tokenizer path in the model
   *    - Create a FastAPI server that wraps the inference pipeline
   * 
   * 2. Server endpoint should accept:
   *    POST /generate_audio
   *    {
   *      "text": "Input text to convert to speech",
   *      "language": "en-ng|ig|yo|ha", 
   *      "speaker": "default"
   *    }
   * 
   * 3. Server should return:
   *    {
   *      "text": "Original or processed text",
   *      "audio_url": "http://server/generated_audio.wav"
   *    }
   * 
   * 4. Replace generateResponse method with:
   *    - HTTP POST to your YarnGPT server
   *    - Handle audio file download/caching
   *    - Error handling for server unavailability
   * 
   * 5. Update this method signature to match real API:
   *    Future<Map<String, dynamic>> generateResponse({
   *      required String text,
   *      required String language,
   *      String speaker = 'default',
   *      String serverUrl = 'http://localhost:8000',
   *    })
   * 
   * Example real implementation:
   * ```dart
   * final response = await http.post(
   *   Uri.parse('$serverUrl/generate_audio'),
   *   headers: {'Content-Type': 'application/json'},
   *   body: jsonEncode({
   *     'text': text,
   *     'language': language,
   *     'speaker': speaker,
   *   }),
   * );
   * return jsonDecode(response.body);
   * ```
   */
}